### PR TITLE
Seat Bidding Backend Functionality and Random Seat Assignment

### PR DESCRIPTION
**Pull request by nowshin-108**
### **Implement Seat Bidding Backend Functionality and Random Seat Assignment**

This PR introduces the backend functionality for the new seat bidding feature and implements a random seat assignment mechanism for flight adding. The seat bidding feature allows passengers to offer their seats for bidding and bid on other available seats within a flight.

**Key Changes:**
1. Added new Express router for bidding operations (/routes/bidding.js)
2. Implemented in-memory storage using Map objects for active biddings and bids
3. Created new API endpoints:
       - POST /bidding/start: Initiate a new bidding
       - POST /bidding/bid: Place a bid on an active bidding
       - POST /bidding/accept: Accept a bid and process seat swap 
       - POST /bidding/cancel: Cancel an active bidding 
       - GET /bidding/active/:flightId: Fetch all active biddings for a flight
4. Integrated WebSocket functionality using Socket.IO for real-time updates
5. Implemented Prisma transactions for atomic seat swaps
6. Added random seat assignment mechanism with attempt loop for conflict resolution
7. Updated database schema to support new functionality

**Schema Updates:**
- Created FlightPassengers table to represent flights and their passengers
- Added Passenger table with userId and seatNumber fields
This allows 
- Easy retrieval of all flights for a specific user (via `Trip`) & all passengers for a specific flight (via `FlightPassengers` and `Passenger`)
- Effective management of seat assignments and prevent duplicates (via unique constraints in `Passenger`)

**Random Seat Assignment:**
Implemented an 'attempt' mechanism to ensure unique seat assignments:
- Generates a random seat number
- Attempts to create a Passenger record with this seat number
- Retries with a new random number if a unique constraint violation occurs
- Limits attempts to 100 to prevent infinite loops
This approach handles concurrent bookings, ensures unique seat assignments, and gracefully manages nearly-full flights.

**New Dependencies:**
- socket.io: For WebSocket communication

**Testing:**
https://www.loom.com/share/d03c6ed4bc43404089a4e40a4959e51d?sid=d78d1c49-5337-43da-825f-836e5a5f57a0

**Documentation:**
https://docs.google.com/document/d/11NAz-kTgxClPSVJnMO0nP5H6LwsQSKEhBmWoWWpJJkg/edit?usp=sharing

**Next Steps:**
- Integrate with frontend seat bidding component

### CODE REVIEW

File: backend/routes/seatbid.js [hunk 1]
```diff
@@ -0,0 +1,186 @@
...
+ */
+router.post('/accept', authenticateUser, async (req, res) => {
+    const { biddingId, bidId } = req.body;
+    const bidding = activeBiddings.get(biddingId);
+
+    if (!bidding || bidding.status !== 'ACTIVE') {
+        return res.status(400).json({ error: 'Bidding not found or not active' });
+    }
+
+    const acceptedBid = bids.get(biddingId).find(bid => bid.id === bidId);
```
Comment (alantulais) [inline backend/routes/seatbid.js:103]: Can we also verify the bid owner still owns their seat?

File: backend/routes/seatbid.js [hunk 2]
```diff
@@ -0,0 +1,186 @@
...
+    if (!acceptedBid) {
+        return res.status(400).json({ error: 'Bid not found' });
+    }
+
+    try {
+        await prisma.$transaction([
+            prisma.passenger.update({
+                where: { flightId_userId: { flightId: bidding.flightId, userId: bidding.passengerId } },
+                data: { seatNumber: 'TEMP' },
```
Comment (alantulais) [inline backend/routes/seatbid.js:112]: what's this? we do we need to assign 'TEMP' and final seat in same transaction?

File: backend/prisma/schema.prisma [hunk 3]
```diff
@@ -41,6 +44,26 @@ model Segment {
...
+model Passenger {
+  passengerId           Int      @id @default(autoincrement())
+  userId                Int
+  user                  User     @relation(fields: [userId], references: [userId])
+  flightId              String
+  flight                Flight   @relation(fields: [flightId], references: [flightId])
+  seatNumber            String
 
+  @@unique([flightId, seatNumber])
+  @@unique([flightId, userId])
```
Comment (alantulais) [inline backend/prisma/schema.prisma:66]: Well implemented.

File: backend/routes/seatbid.js [hunk 4]
```diff
@@ -0,0 +1,186 @@
...
+    if (!bidding || bidding.status !== 'ACTIVE') {
+        return res.status(400).json({ error: 'Bidding not found or not active' });
+    }
+
+    const bidderActiveBidding = Array.from(activeBiddings.values()).find(
+        b => b.passengerId === bidderId && b.status === 'ACTIVE'
+    );
+
+    if (bidderActiveBidding) {
+        bidderActiveBidding.status = 'CANCELLED';
```
Comment (alantulais) [inline backend/routes/seatbid.js:70]: So we'll automatically cancel previous biddings while bidding in a new auction. 

I would prefer us to allow users to bid in multiple auctions till they still own their seat. Feel free to do that in follow up PR.

**Review by alantulais — APPROVED** [review]
Overall LGTM, good job @nowshin-108.

I left a few questions/suggestions, but nothing concerning about this diff. My major ask is to allow users to bid in multiple biddings in parallel, as long as they haven't swapped their seat already. 

Feel free to address that in separate PR.

**Review by alantulais — APPROVED** [review]

File: backend/routes/seatbid.js (hunk 1 above)
Comment (nowshin-108) [reply backend/routes/seatbid.js:103]: added in the next PR

File: backend/routes/seatbid.js (hunk 4 above)
Comment (nowshin-108) [reply backend/routes/seatbid.js:70]: added in the next PR

File: backend/routes/seatbid.js (hunk 2 above)
Comment (nowshin-108) [reply backend/routes/seatbid.js:112]: My database has unique constraint on seat number and I can't directly swap seats which violated the unique constraint momentarily. So I used the 'TEMP' seat assignment as an intermediate step in the transaction to temporarily free up the initiator's seat. It allows a clean swap of seat numbers between two passengers without violating the unique constraint on seat assignments for a given flight.

**Pull request merged**